### ISSUE TITLE
[Work group] Add RLHF like ColosallChat on bigger dataset to achieve ChatGPT quality

### ISSUE THREAD

**Issue by alyxdow** [ggerganov/llama.cpp#743]
# Add RLHF like ColosallChat on bigger dataset to achieve ChatGPT quality
Although models in the GPT series, such as ChatGPT and GPT-4, are highly powerful, they are unlikely to be fully open-sourced. Fortunately, the open-source community has been working hard to address this.

For example, Meta has open-sourced the LLaMA model, which offers parameter sizes ranging from 7 billion to 65 billion. A 13 billion parameter model can outperform the 175 billion GPT-3 model on most benchmark tests. However, since it doesn’t have an instruct tuning stage, its actual generated results are not satisfactory.

Stanford’s Alpaca generates training data in a self-instructed manner by calling OpenAI’s API. With only 7 billion parameters, this lightweight model can be fine-tuned at a fraction of the cost to achieve conversational performance similar to a very large language model like GPT-3.5 with 175 billion parameters.

However, existing open-source solutions can only be considered as supervised fine-tuned models in the first stage of RLHF (Reinforcement Learning from Human Feedback), with subsequent alignment and fine-tuning stages not performed. Additionally, Alpaca’s training dataset is limited to English, which to some extent restricts the model’s performance.

Yet, the impressive effects of ChatGPT and GPT-4 are due to the introduction of RLHF into the training process, which increases the consistency of the generated content with human values.

![image](https://user-images.githubusercontent.com/84633629/229590135-fc856c3f-b138-4fad-842e-e0bf34ffd69f.png)

**Comment by alyxdow** [comment]
[OIG Dataset](https://huggingface.co/datasets/laion/OIG)

I came across the OIG dataset, which is a large open-source instruction dataset consisting of approximately 43M instructions. I believe this dataset could be a valuable addition to GitHub for those working on chatbot technology and related projects.

The dataset has been released by LAION.ai, along with its volunteers, Ontocord, Together and other members of the open source community. The purpose of this release is to create equal access to chatbot technology and encourage improvements from contributors.

I would like to suggest including the OIG dataset on GitHub to make it easily accessible to the community and allow others to contribute to its development.

**Comment by alyxdow** [comment]
**As we work towards training the RHLF LLaMa model, I propose that we stack all the datasets we have found for this task. By combining these datasets, we can increase the variety and size of our training data, which will ultimately improve the accuracy and performance of our model.

I suggest that we carefully review each dataset to ensure that they are relevant to our task and that they meet our quality standards. We should also consider the format and structure of each dataset and determine the best way to combine them into a single training set.**

**Comment by alyxdow** [comment]
![lama-alpaca](https://user-images.githubusercontent.com/84633629/229592183-b927ca5e-a60d-45fa-a142-99d87e01027e.gif)

**Comment by prusnak** [comment]
How is any of this related to llama.cpp?

**Comment by Royalphax** [comment]
Not an expert but it seems to be very well explained. I think that the question of how we can improve the quality of results we get during inference by using RLHF algorithm on datasets as you described is very interesting.

Now, is it the role of the llama.cpp project?
In my opinion : not sure.

I think llama.cpp is here to allow us to convert, quantize and mostly run (inference phase) models in the most optimized way.

Maybe the question of how to optimize current datasets with RLHF should be the goal of another open-source project. And/or this conversation should be moved to the "discussion" tab.

**Comment by jon-chuang** [comment]
Similar approach to RLHF (but lower quality) is used to train alpaca and vicuna, among others. Vote to close this issue.
